pkg/prometheus/operator.go

```diff
@@ -2179,3 +2198,20 @@ func validateProberURL(url string) error {
 	}
 	return nil
 }
+
+func validateScrapeIntervalAndTimeout(p *monitoringv1.Prometheus, scrapeInterval, scrapeTimeout string) error {
+	if scrapeInterval != "" && scrapeTimeout != "" {
+		return operator.CompareScrapeTimeoutToScrapeInterval(scrapeTimeout, scrapeInterval)
+	}
+
+	if scrapeInterval != "" && scrapeTimeout == "" {
+		if p.Spec.ScrapeTimeout != "" {
+			return operator.CompareScrapeTimeoutToScrapeInterval(p.Spec.ScrapeTimeout, scrapeInterval)
+		}
+	}
+
+	if scrapeInterval == "" && scrapeTimeout != "" {
+		return operator.CompareScrapeTimeoutToScrapeInterval(scrapeTimeout, p.Spec.ScrapeInterval)
+	}
+	return nil
```

Comment:
I think that it can be simplified a bit. If the scrape timeout isn't defined (or 0) then there's nothing to check because Prometheus will do the right thing (if scrape interval > global timeout then timeout = scrape interval otherwise timeout = global timeout).

```suggestion
	if scrapeTimeout == "" {
		return nil
	}

	if scrapeInterval == "" {
		scrapeInterval = p.Spec.ScrapeInterval
		if scrapeInterval == "" {
			scrapeInterval = "30s"
		}
	}

	return operator.CompareScrapeTimeoutToScrapeInterval(scrapeTimeout, scrapeInterval)
```

Reply:
We need this part in `CompareScrapeTimeoutToScrapeInterval()` since we use that in prometheus crd level validation as well https://github.com/prometheus-operator/prometheus-operator/pull/4491/files#diff-453ce9c721c51bf9591c2b212e9ce50157b9c40d1b211177bf502be268099656R254-R256

```go
if scrapeInterval == "" {
   scrapeInterval = "30s"
}
```

Reply:
@simonpasquier I updated per above above suggestion but this wont handle 2nd case mentioned right?. 

`if the endpoint defines only the scrape interval value then compare it with the global scrape timeout (nothing to do if unset).`

https://github.com/prometheus-operator/prometheus-operator/pull/4491#pullrequestreview-851515889

Reply:
Reviewed it here so we are good https://github.com/prometheus/prometheus/blob/2c61d29b2a865db7e1a8d188e08b68ed154929c8/config/config.go#L292-L301

Reply:
Unless I am missing something, when I read the API docs it isn't clear to me that the endpoint config, if not specified will be inherited from the Prometheus CR or that if one field is provided it will be compared against the global field so maybe we should call that out?

That feels slightly obtuse in the "self-service" model however. Perhaps we could take an all or nothing approach and declare than unless both fields are required to be either omitted or provided as a pair, but that would be a breaking change i guess.

Reply:
From what I understand thats default set by prometheus when scrape interval not specified in job section it sets to global scrape interval and thats not handled by operator https://github.com/prometheus/prometheus/blob/2c61d29b2a865db7e1a8d188e08b68ed154929c8/config/config.go#L289-L290

Reply:
I can update that in separate pr